src/format/OpData01.cpp

```diff
@@ -0,0 +1,143 @@
+    QByteArray::const_iterator it_end = data.cend();
+    if (!data.startsWith(header)) {
+        RETURN_OPDATA01_ERROR_INSTANCE(QString("data claims it does not start with header: %1")
+                                           .arg(QString::fromUtf8(data.mid(0, header.size()).toHex())));
+    } else {
+        it += header.size();
+    }
+
+    qlonglong qlen(0LL);
+    int len;
+    {
+        for (int i = 0; i < 8 and it != it_end; ++i, ++it) {
+            auto ch = static_cast<unsigned char>((*it) & 0xFF);
+            qlen = (ch << (8 * i)) | qlen;
+        }
+
+        // stop kidding ourselves and having to downcast len
+        len = int(qlen);
+    }
+    // qDebug("toLongLong := %lld\n", qlen);
+
+    QByteArray iv(16, '\0');
+    for (int i = 0, iv_len = 16; i < iv_len and it != it_end; ++i, ++it) {
+        char ch = *it;
+        iv[i] = ch;
+    }
+    if (16 != iv.size()) {
+        RETURN_OPDATA01_ERROR_INSTANCE(QString("short IV read, wanted %1 but got %2").arg(16).arg(iv.size()));
+    }
+
+    SymmetricCipher cipher(SymmetricCipher::Aes256, SymmetricCipher::Cbc, SymmetricCipher::Decrypt);
+    if (!cipher.init(key, iv)) {
+        RETURN_OPDATA01_ERROR_INSTANCE(QString("Unable to init cipher for opdata01: %1").arg(cipher.errorString()));
+    }
+
+    QByteArray* clearText;
+    {
+        const int blockSize = cipher.blockSize();
+        // qDebug("cipher.blockSize = %d\n", blockSize);
+
+        int randomBytes = blockSize - (len % blockSize);
+        // qDebug("cipher.randomBytes1 = %d\n", randomBytes);
+        if (0 == randomBytes) {
+            // add random blk
+            randomBytes = blockSize;
+        }
+        int clear_len = len + randomBytes;
+        // qDebug("cipher.randomBytes2 = %d\tclearText.len=%d\n", randomBytes, clear_len);
+        QByteArray qbaCT(clear_len, '\0');
+        for (int i = 0; i < clear_len and it != it_end; ++i, ++it) {
+            qbaCT[i] = *it;
+        }
+
+        const int hmacLen = 256 / 8;
+        QByteArray hmacSig; // 256 / 8, '\0');
+        for (int i = 0; i < hmacLen and it != it_end; ++i, ++it) {
+            hmacSig.append(*it);
+        }
+        if (hmacSig.size() != hmacLen) {
+            RETURN_OPDATA01_ERROR_INSTANCE(
+                QString("Short HMAC sig read; wanted %1 but got %2").arg(hmacLen).arg(hmacSig.size()));
+        }
+
+        if (it < it_end) {
+            qWarning("Did not consume all of opdata01!");
+        } else if (it > it_end) {
+            qWarning("opdata01 iterator greater than const_end");
+        }
+
+        const QByteArray actualHmac =
+            CryptoHash::hmac(data.mid(0, data.size() - hmacSig.size()), hmacKey, CryptoHash::Algorithm::Sha256);
+        if (actualHmac != hmacSig) {
+            RETURN_OPDATA01_ERROR_INSTANCE(QString("opdata01 failed its HMAC; wanted %1 but got %2")
+                                               .arg(QString::fromUtf8(hmacSig.toHex(':')))
```

Comment:
QByteArray::toHex with a separator was [introduced in Qt 5.9](http://doc.qt.io/qt-5/qbytearray.html#toHex-1). I believe the minimum supported version (per INSTALL.md) is Qt 5.2. There are additional calls to this function in this PR.

Reply:
> I believe the minimum supported version (per INSTALL.md) is Qt 5.2

Related to that, the state of `develop` (and `2.3.4`) against 5.2 is in rough shape.

[`QAtomicInteger`](http://doc.qt.io/qt-5.6/qatomicinteger.html) was introduced in 5.3 but is used by [`src/browser/NativeMessagingBase.h`](https://github.com/keepassxreboot/keepassxc/blob/2.3.4/src/browser/NativeMessagingBase.h#L31). Replacing it with `QAtomicInt` makes the file at least compile, but then it surfaces this next:

```
/opt/qt52/include/QtCore/qjsonvalue.h: In member function 'QJsonObject BrowserAction::handleSetLogin(const QJsonObject&, const QString&)':
/opt/qt52/include/QtCore/qjsonvalue.h:119:12: error: 'QJsonValue::QJsonValue(const void*)' is private
     inline QJsonValue(const void *) {}
            ^
/keepassxc/src/browser/BrowserAction.cpp:332:22: error: within this context
     message["error"] = "";
```
Wrapping those literals in `QString()` gets us past that, but then:

```
/keepassxc/src/browser/BrowserService.cpp: In member function 'QJsonArray BrowserService::findMatchingEntries(const QString&, const QString&, const QString&, const QString&, const StringPairList&)':
/keepassxc/src/browser/BrowserService.cpp:265:16: error: no match for 'operator<<' (operand types are 'QJsonArray' and 'QJsonObject')
         result << prepareEntry(entry);
                ^
/keepassxc/src/browser/BrowserService.cpp:265:16: note: candidates are:
In file included from /opt/qt52/include/QtCore/qstring.h:45:0,
                 from /opt/qt52/include/QtCore/qjsonvalue.h:46,
                 from /opt/qt52/include/QtCore/qjsonarray.h:45,
                 from /opt/qt52/include/QtCore/QJsonArray:1,
                 from /keepassxc/src/browser/BrowserService.cpp:20:
/opt/qt52/include/QtCore/qchar.h:521:28: note: QDataStream& operator<<(QDataStream&, QChar)
```
changing to `.append()` cures that

---

One can see this for themselves via:
```diff
diff --git a/Dockerfile b/Dockerfile
index e1fd26d4..89a30289 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -18,8 +18,8 @@ FROM ubuntu:14.04
 
 ENV REBUILD_COUNTER=10
 
-ENV QT5_VERSION=qt510
-ENV QT5_PPA_VERSION=qt-5.10.1
+ENV QT5_VERSION=qt52
+ENV QT5_PPA_VERSION=qt521
 ENV TERM=xterm-256color
 
 RUN set -x \
@@ -55,6 +55,7 @@ RUN set -x \
         libxi-dev \
         libxtst-dev \
         mesa-common-dev \
+        libgl1-mesa-dev \
         libyubikey-dev \
         libykpers-1-dev \
         libqrencode-dev \
```
(which kind of surprises me isn't the default build configuration, given the stated minimum dep)

I can create a PR against `develop` to cure those, if you'd like, but I wanted to make you aware.